File: 2020/day15/solve.py

def play(numbers, turns):
    last_turns = {n: [i] for i, n in enumerate(numbers)}
    last_spoken = numbers[-1]
    for t in range(len(numbers), turns):
        if len(last_turns[last_spoken]) == 1:
            spoken = 0
        else:
            spoken = last_turns[last_spoken][-1] - last_turns[last_spoken][-2]
        last_spoken = spoken

        if spoken in last_turns:
            last_turns[spoken] = [last_turns[spoken][-1], t]
        else:
            last_turns[spoken] = [t]
    return last_spoken


def main():
    with open('input.txt') as f:
        numbers = [int(x) for x in f.read().split(',')]

    print(play(numbers, 2020))
    print(play(numbers, 30000000))


if __name__ == '__main__':
    main()

File: 2020/day16/solve.clj
(require 'clojure.java.io)
(require 'clojure.string)

(def debug true)
(def lines
  (with-open [rdr (clojure.java.io/reader (if debug "small.txt" "input.txt"))]
    (reduce conj [] (line-seq rdr))))

(def specs (reduce
  #(if (= %2 "")
    (conj %1 [])
    (cons (cons %2 (first %1)) (rest %1)))
  [] lines))

(def intervals
  (letfn [(parse-field-spec [s]
            (let [parts (clojure.string/split s #" ")
                  field-name (first parts)
                  intervals (map #() (range 1 (count parts) 2))
                  ]

              ))]
    (map parse-field-spec (first specs))))
